shared/src/clients/whatsapp-client.class/index.ts

```diff
@@ -0,0 +1,156 @@
+        }
+    }
+  }
+  public async validateSingleRecipient(
+    input: WhatsAppTemplateMessageToSend,
+    isLocal = false
+  ): Promise<WhatsAppId> {
+    // bypass validation in local environment because no access to on-prem API
+    // proxy can only send message, no endpoint for validation
+    if (isLocal) {
+      return input.to
+    }
+    const { token, url } = this.getCredentials(input.apiClient)
+    const res = await this.axiosInstance
+      .request<ValidateContact200Response>({
+        method: 'post',
+        url: CONTACT_ENDPOINT,
+        baseURL: url,
+        headers: { Authorization: `Bearer ${token}` },
+        data: { blocking: 'wait', contacts: [input.to] },
+      })
+      .catch((err: Error | AxiosError) => {
+        if (axios.isAxiosError(err) && err.response) {
+          throw new Error(
+            `Error validating recipient ${input.to}: ${JSON.stringify(
+              err.response
+            )}`
+          )
+        }
+        throw new Error(
+          `Unexpected error validating recipient ${
+            input.to
+          }. Error: ${JSON.stringify(err)}`
+        )
+      })
+    const {
+      data: { contacts },
+    } = res
+    return contacts[0].wa_id as WhatsAppId
+  }
+
+  // public async validateMultipleRecipients(
+  //   input: WhatsAppTemplateMessageToSend[]
+  // ): Promise<WhatsAppId[]> {
+  //   const { tokenOne, urlOne } = this.getCredentials(
+  //     WhatsAppApiClient.clientOne
+  //   )
+  //   const { tokenTwo, urlTwo } = this.getCredentials(
+  //     WhatsAppApiClient.clientTwo
+  //   )
+  // }
+
+  public async sendMessage(
+    input: WhatsAppTemplateMessageToSend,
+    isLocal = false
+  ): Promise<MessageId> {
+    const { token, url } = this.getCredentials(input.apiClient)
+    // modify headers and baseURL if local environment to send to proxy
+    const { headers, baseURL } = isLocal
+      ? {
+          headers: {
+            Authorization: `Bearer proxy=${this.credentials['proxyToken']} client=${token}`,
+          },
+          baseURL: `${this.credentials['proxyUrl']}`,
+        }
+      : { headers: { Authorization: `Bearer ${token}` }, baseURL: url }
+    const {
+      data: { messages },
+    } = await this.axiosInstance
+      .request<TemplateMessage200Response>({
+        method: 'post',
+        url: MESSAGE_ENDPOINT,
+        baseURL,
+        headers,
+        data: {
+          to: input.to,
+          type: 'template',
+          template: {
+            namespace: this.credentials['namespace'],
+            name: input.templateName,
+            language: {
+              code: input.language,
+              policy: 'deterministic',
+            },
+          },
+          components: [],
+        },
+      })
+      .catch((err: Error | AxiosError) => {
+        if (
+          axios.isAxiosError<TemplateMessageErrResponse>(err) &&
+          err.response
+        ) {
+          const { errors } = err.response.data
+          throw new Error(
```

Comment:
Can we created error types to differentiate errors that are up for downstream users to handle?

Reply:
done